### PR TITLE
add DeepHPMs examples and FunctionalLoss&FunctionalMetric apis

### PR DESCRIPTION
**Pull request by lijialin03**
### PR types

Others
### PR changes

Others
### Describe

add 5 DeepHPMs examples
add ppsci.loss.FunctionalLoss api
add ppsci.metric.FunctionalMetric api
update docs(zh)

### CODE REVIEW

File: examples/deephpms/burgers.py [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+        eval_during_train=False,
+        validator=validator_pde,
+    )
+
+    # train model
+    solver.train()
+    # evaluate after finished training
+    solver.eval()
+
+    # model 3: solution net
```
Comment (HydrogenSulfate) [inline examples/deephpms/burgers.py:250]: `# stage 3: training solution net`

File: examples/deephpms/burgers.py [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+    # update transform of model 2 to update calculation of loss
```
Comment (HydrogenSulfate) [inline examples/deephpms/burgers.py:251]: `# re-register transform for model 2, fit for loss of stage 3`

File: examples/deephpms/burgers.py [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+    solver = ppsci.solver.Solver(
+        model=solver.model,
+        constraint=constraint_sol,
+        output_dir=OUTPUT_DIR,
+        optimizer=optimizer_sol,
+        epochs=EPOCHS,
+        iters_per_epoch=ITERS_PER_EPOCH,
+        eval_during_train=False,
+        validator=validator_sol,
+    )
```
Comment (HydrogenSulfate) [inline examples/deephpms/burgers.py:343]:
``` py
solver = ppsci.solver.Solver(
    solver.model,
    constraint_sol,
    OUTPUT_DIR,
    optimizer_sol,
    None,
    EPOCHS,
    ITERS_PER_EPOCH,
    eval_during_train=False,
    validator=validator_sol,
)
```

**Review by HydrogenSulfate — CHANGES_REQUESTED** [review]

File: ppsci/metric/func.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        >>> def metric_expr(output_dict):
+        ...     rel_l2 = paddle.norm(output_dict - output_dict) / paddle.norm(output_dict)
+        ...     return {"l2": rel_l2}
+        >>> metric_dict = ppsci.metric.FunctionalMetric(metric_expr)
+    """
+
+    def __init__(
+        self,
+        metric_expr: Optional[Callable] = None,
+    ):
```
Comment (HydrogenSulfate) [inline ppsci/metric/func.py:40]: 新增了一个参数 `keep_batch`，参考
https://github.com/PaddlePaddle/PaddleScience/blob/release%2F1.0/ppsci/metric/base.py#L21
修改下

File: ppsci/metric/func.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        super().__init__()
+        self.metric_expr = metric_expr
+
+    def forward(self, output_dict, label_dict=None, weight_dict=None):
```
Comment (HydrogenSulfate) [inline ppsci/metric/func.py:44]: metric 是没有 weight_dict的，删掉

**Review by HydrogenSulfate — CHANGES_REQUESTED** [review]
案例的修改参考第一、二个案例的修改建议，其余的按照review建议修改一下

File: examples/deephpms/burgers.py [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+# limitations under the License.
+
+import paddle
+import paddle.nn.functional as F
+
+import ppsci
+from ppsci.utils import logger
+
+
+def pde_loss_compute(output_dict):
```
Comment (HydrogenSulfate) [inline examples/deephpms/burgers.py:22]: `_compute` 改为 `_func`

**Review by HydrogenSulfate — CHANGES_REQUESTED** [review]

File: ppsci/loss/func.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+from typing import Optional
+from typing import Union
+
+from typing_extensions import Literal
+
+from ppsci.loss import base
+
+
+class FunctionalLoss(base.Loss):
+    r"""Class for functional loss.
```
Comment (HydrogenSulfate) [inline ppsci/loss/func.py:26]: 细化一下class说明：`Functional loss class, which allows to use custom loss computing function from given loss_expr for complex computation cases.`

File: ppsci/metric/func.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+# limitations under the License.
+
+from typing import Callable
+from typing import Optional
+
+from ppsci.metric import base
+
+
+class FunctionalMetric(base.Metric):
+    r"""Class for functional metric.
```
Comment (HydrogenSulfate) [inline ppsci/metric/func.py:22]: 细化一下class说明：Functional metric class, which allows to use custom mtric computing function from given metric_expr for complex computation cases.

**Review by HydrogenSulfate — CHANGES_REQUESTED** [review]

**Review by HydrogenSulfate — APPROVED** [review]
LGTM
